File: 1302-deepest-leaves-sum/1302-deepest-leaves-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        
        s = 0
        h = float('-inf')
        
        def dfs(root, curr_height):
            nonlocal s, h;
            if root:
                if not root.left and not root.right:
                    if curr_height > h:
                        s = root.val
                        h = curr_height
                    elif curr_height == h:
                        s += root.val
                dfs(root.left, curr_height + 1)
                dfs(root.right, curr_height + 1)
                
        dfs(root, 0)
        
        return s
        

File: 1302-deepest-leaves-sum/1302-deepest-leaves-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    public:    
    int x = 0;
    int depth;
    
    int deepestLeavesSum(TreeNode* root) 
    {
        depth = 1;
        helper(root, 1);
        return x;
    }
    
    void helper(TreeNode* root, int level)
    {
        if(root == NULL)
            return;
        if(level == depth)
            x += root->val;
        else if(level > depth)
        {
            x = root->val;
            depth = level;
        }
        helper(root->left, level + 1);
        helper(root->right, level + 1);
    }
};